gpkit/constraint_meta.py

```diff
@@ -0,0 +1,123 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
+
+
+class Constraint(object):
+    __metaclass__ = ABCMeta
+
+    @abstractproperty
+    def varkeys(self):
+        "Varkeys present in the Constraint"
+        pass
+
+    @abstractproperty
+    def substitutions(self):
+        "Substitutions that the Constraint currently knows about"
+        pass
+
+    @substitutions.setter
+    def substitutions(self, newvalue):
+        ".substitutions mist be settable"
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        "string representation of the constraint, used for printing programs"
+        pass
+
+    @abstractmethod
+    def latex(self):
+        "latex representation of the constraint, used for printing programs"
+        pass
+
+    def _repr_latex_(self):
+        "Classes that inherit from Constraint will show as latex in IPython"
+        return "$$" + self.latex() + "$$"
+
+    @abstractmethod
+    def sub(self, subs, value=None):
+        "Returns new object without any of the varkeys in subs.keys()"
+        pass
+
+    @abstractmethod
+    def process_result(self, result):
+        """Does arbitrary computation / manipulation of a program's result
+
+        There's no guarantee what order different constraints will process
+        results in, so any changes made to the program's result should be
+        careful not to step on other constraint's toes.
+
+        Potential Uses
+        --------------
+          - check that an inequality was tight
+          - add values computed from solved variables
+
+        """
+        pass
+
+
+class GPConstraint(Constraint):
+    "Interface for constraints representable as posynomials <= 1"
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def as_posyslt1(self):
+        "Returns list of posynomials which must be kept <= 1"
+        pass
+
+    @abstractmethod
+    def sens_from_dual(self, p_senss, m_sensss):
+        """Computes constraint and variable sensitivities from dual solution
+
+        Arguments
+        ---------
+        p_senss : list
+            Sensitivity of each posynomial returned by `self.as_posyslt1()`
+
+        m_sensss: list of lists
+            Each posynomial's monomial sensitivities
+
+
+        Returns
+        -------
+        constraint_sens : dict
+            The interesting and computable sensitivities of this constraint
```

Comment:
For example, equality constraints have interesting sensitivities, while for regular `m >= p` constraints there's no need to give both the sensitivity of the whole thing and the sensitivity of `m`.

For truss problems, the shadow prices can be turned into geometric gradients, but only under the right organizing framework (i.e., one constraint "set" for every beam).

I think that writeup will be good: what I'll particularly try to emphasize is that sensitivities reward thought.